File: 0008-string-to-integer-atoi/0008-string-to-integer-atoi.cpp
class Solution {
public:
    int myAtoi(string s) {
        // int n = s.length();
        
        // int positive = 0;
        // int negative = 0;
        // int i = 0;

        // while(s[i] == ' ')
        // {
        //     i++;
        // }

        // while(s[i] == '+')
        // {
        //     positive++;
        // }

        // while(s[i] == '-')
        // {
        //     negative++;
        // }

        // while(i < n)

        int i = 0;
  while (i < s.length() && isspace(s[i])) {
    i++;
  }

  // Check for sign.
  int sign = 1;
  if (i < s.length() && (s[i] == '+' || s[i] == '-')) {
    sign = (s[i] == '-') ? -1 : 1;
    i++;
  }

  // Read digits.
  int result = 0;
  while (i < s.length() && isdigit(s[i])) {
    int digit = s[i] - '0';
    if (result > (INT_MAX - digit) / 10) {
      return sign == 1 ? INT_MAX : INT_MIN;
    }
    result = result * 10 + digit;
    i++;
  }

  // Apply sign and return result.
  return sign * result;


    }
};

File: 0008-string-to-integer-atoi/0008-string-to-integer-atoi.java
class Solution {
    public int myAtoi(String s) {

        int i = 0;
        int n = s.length();

      

        while(i < n && s.charAt(i) == ' ')
        {
            i++;
        }

        if(i == n)
        {
            return 0;
        }

        char sign = '+';

       
            if(s.charAt(i) == '+')
            {
                sign = '+';
                i++;
            }
            else if(s.charAt(i) == '-')
            {
                sign = '-';
                i++;
            }
            else if(!Character.isDigit(s.charAt(i)))
            {
                return 0;
            }

            while(i < n && s.charAt(i) == '0')
            {
                i++;
            }

            int num = 0;

            while(i < n && Character.isDigit(s.charAt(i)))
            {
                int val = (int)(s.charAt(i) - '0');

                if(num > (Integer.MAX_VALUE/10) ||  ( ( num == (Integer.MAX_VALUE/10) ) && (val > (Integer.MAX_VALUE%10)) ) )
                {
                    return sign == '+' ? Integer.MAX_VALUE : Integer.MIN_VALUE;
                }

                num = (num*10) + val;
                i++;
            }

            return sign == '+' ? num : (-1*num);
        
        
    }
}